aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java

```diff
@@ -360,4 +421,22 @@ protected void finalize() throws Throwable {
       LOG.warn("Unclosed output stream created by:\n\t{}", trace);
     }
   }
+
+  private static class FileWithEtag {
+    private final File file;
+    private final MessageDigest eTag;
```

Comment:
Why keep the `MessageDigest` around when the only use of it is to immediately call `digest()` on line 269? Seems like this should probably track the digest bytes with the file instead.

Reply:
I think we can do that, it will add the logic a tad bit more complicated as we close stream at multiple places. Will update it.